src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,862 @@
+                    else
+                    {
+                        // Creating a SyntaxToken from the new parameter name and adding an annotation to it
+                        // and passing that syntaxtoken in to generator.IdentifierName to create a new
+                        // IdentifierNameSyntax node.
+                        // Need to create the RenameAnnotation on the token itself otherwise it does not show up
+                        // properly.
+                        replacement = (TIdentifierNameSyntax)generator.IdentifierName(generator.Identifier(parameterName)
+                            .WithAdditionalAnnotations(RenameAnnotation.Create()));
+                        editor.ReplaceNode(match, replacement);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Goes through the parameters of the original method to get the location that the parameter
+        /// and argument should be introduced.
+        /// </summary>
+        private static int GetInsertionIndex(Compilation compilation, IMethodSymbol methodSymbol,
+            ISyntaxFactsService syntaxFacts, SyntaxNode methodDeclaration)
+        {
+            var parameterList = syntaxFacts.GetParameterList(methodDeclaration);
+            Contract.ThrowIfNull(parameterList);
+            var insertionIndex = 0;
+
+            foreach (var parameterSymbol in methodSymbol.Parameters)
+            {
+                // Want to skip optional parameters, params parameters, and CancellationToken since they should be at
+                // the end of the list.
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.IsParams &&
+                        !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken)?.FullName!)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument.
+        /// </summary>
+        private async Task<Solution> RewriteSolutionAsync(
+            Document originalDocument, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            SyntaxNode containingMethod, bool allOccurrences, string parameterName,
+            ImmutableDictionary<Document, List<TInvocationExpressionSyntax>> callSites, bool trampoline,
+            bool overload, CancellationToken cancellationToken)
+        {
+            var modifiedSolution = originalDocument.Project.Solution;
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
```

Comment:
this is not appropriate.  not all descendant nodes are varialbles.  for example `a.b.c + x.y.z`.  Only `a` and `x` are potental variables, `b, c, y, z` are all identifier names, but are not variables.

Reply:
In that case, should I just iterate through the keys of the mappingDictionary? Because that iterates through all of the identifiers and ties them back to parameters, if they can be tied back